ui/app/pages/confirmation/confirmation.js

```diff
@@ -0,0 +1,243 @@
+import React, {
+  useCallback,
+  useEffect,
+  useMemo,
+  useReducer,
+  useState,
+} from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { useHistory } from 'react-router-dom';
+import { isEqual } from 'lodash';
+import { produce } from 'immer';
+import { getEnvironmentType } from '../../../../app/scripts/lib/util';
+import {
+  ENVIRONMENT_TYPE_FULLSCREEN,
+  ENVIRONMENT_TYPE_POPUP,
+} from '../../../../shared/constants/app';
+import ConfirmationFooter from '../../components/app/confirmation-footer';
+import Box from '../../components/ui/box';
+import Chip from '../../components/ui/chip';
+import MetaMaskTemplateRenderer from '../../components/app/metamask-template-renderer';
+import SiteIcon from '../../components/ui/site-icon';
+import { DEFAULT_ROUTE } from '../../helpers/constants/routes';
+import { stripHttpSchemes } from '../../helpers/utils/util';
+import { useI18nContext } from '../../hooks/useI18nContext';
+import { useOriginMetadata } from '../../hooks/useOriginMetadata';
+import { getUnapprovedConfirmations } from '../../selectors';
+import NetworkDisplay from '../../components/app/network-display/network-display';
+import { COLORS, SIZES } from '../../helpers/constants/design-system';
+import Callout from '../../components/ui/callout';
+import { getTemplateValues, getTemplateAlerts } from './templates';
+
+/**
+ * a very simple reducer using produce from Immer to keep state manipulation
+ * immutable and painless. This state is not stored in redux state because it
```

Comment:
This seems like a curious decision :thinking: We already have state in Redux that is applicable only to one session and page. I don't think that's a reason to keep it _out_ of Redux. Using a store would make this easier to debug as well. And might improve postmortem debugging when/if we add Redux breadcrumbs to our Sentry reports.

Reply:
It seems to be overkill, to me, to place something in redux that will only ever be used in a single view. Redux should be for the application-wide state, as per the redux-toolkit style guide recommendation [evaluate where each piece of state should live](https://redux.js.org/style-guide/style-guide#evaluate-where-each-piece-of-state-should-live)

Reply:
Yeah, I guess. But, is this less complex than putting it in Redux? I saw that section as talking about when to use component state v.s. Redux, because in general keeping state in components is simpler if it concerns only those components. But creating intermediary reducers seems like _a lot_.

Maybe not though. I'll take another read through - I don't understand the context of this very well yet.

Reply:
It actually started out as a use state, but the update logic was messy requiring a bunch of object spreads. Immer was a spur of the moment choice but it resulted in code that could very well be moved back to a useState hook instead of useReducer. Three things made me think to put in an intermediary useReducer instead of useState/redux

1. The state of this page will not be needed anywhere else. The state on this page is determined by the templates that are rendered, and would have to be loaded into redux on each new confirmation only for it to be immediately consumed at the page level.
2. I needed an object which React recommends breaking multiple keys into multiple useState but that doesn't work with the programmatic nature of this state.
3. It fits the paradigm of a reducer and the update logic was messy in a setState call before immer

Reply:
The reason I consider this as "application-wide state", because the alert state is interwoven with _a lot_ of other things, such as the confirmation properties, selected address, selected network, eth balance, token balance, etc. I see that this isn't handled in this reducer right now, but rather as part of the template `getAlerts` function. But this is exactly the sort of thing Redux is well suited for handling.

This reducer is simple and lacking any interaction with application-wide state at the moment only because the complex state logic is in `getAlerts`.

Reply:
It was _intentional_ to isolate it from redux, though, because third-party developers in the future will presumably not have access to the redux store. If we did just provide a filtered state tree from redux to the `getAlerts` method provided by third parties, it seems safer to isolate the returned state from the rest of our application state.

I would contest that simply accessing application-wide state is not enough of a signal to warrant any derived state also presiding in the application-wide state.

I am not really following what you're suggesting here. Are you suggesting removing/significantly rework the `getAlerts` API for the template system?

Reply:
Right, I hadn't considered third-party templates. I was thinking of this as another of our pages.

Where I was coming from with this is that I've been trying to ensure we keep our front-end components as simple as possible and move UI application logic into Redux where possible. I'd even say that too much logic in our components has historically been the a big part of _most_ of the technical debt in this project. So I'm having trouble envisioning how we'll keep that from happening again with these templates as the UI managed with templates gets increasingly complex.

Reply:
Well, it definitely isn't my intention to _create technical debt_. I think that we should probably define some of these concepts in a formal, written, and unanimously approved engineering principles/design document that mirrors the redux toolkit one. In our document I'd love for us to have a more clear definition of complexity as it relates to [move complex logic outside components](https://redux.js.org/style-guide/style-guide#move-complex-logic-outside-components). Having a hive mind about how we approach implementation as much as possible is going to be super important as we rapidly scale our team. 

For what it's worth I do see that it's impossible to not cross the line of complexity as more confirmations move to use the template system. I tackled this project with the concept of having this work for third-party consumers, without thinking about how we could best utilize the same overall system.

I do think that as we tackle the send confirmation or any of the others that the use cases for the state within the confirmation page are going to expand. I'd like to take time to think of a store structure that would allow many permutations of state for the confirmation page and implement it as part of the next confirmation that gets worked into this system.

Would that be agreeable to you, to address this concern as I work through the next confirmation implementation?

Reply:
Absolutely, that sounds like a good plan :+1: